sympy/vector/scalar.py

```diff
@@ -69,3 +69,7 @@ def system(self):
 
     def _sympystr(self, printer):
         return self._name
+
+    def func(self, *args):
+        return self.__new__(self.__class__, *args, pretty_str =
+                self._pretty_form[1:], latex_str = self._latex_form[2:])
```

Comment:
Something very strange here (on master):
```python
In [33]: from sympy.vector import CoordSys3D

In [34]: N = CoordSys3D('N')

In [35]: s = N.x

In [36]: s
Out[36]: x_N

In [37]: s.func(*s.args)
Out[37]: x0

In [38]: s.func(*s.args) == s
Out[38]: True
```
Is it nos possible to change it so that the optional arguments for `BaseScalar` are passed to `super().__new__` and appear in args?

Reply:
Same question applies to `BaseVector`.

Reply:
Currently the names are stored as `str`, which is why they are not in `args`. Storing them in a `String` wrapper is possible, but it would change the api (although only in `_` attributes). Note that this is currently an `AtomicExpr`, so its behaviour might be unexpected when there is an update for `func` of `Atom`. Redesigning these classes to simply not be `Atom` subclasses might be the best approach, but I felt that might be a bit invasive.

Reply:
Can we not just store `String` in `_args` (i.e. by passing to `super().__new__`) even if it is redundant. It's usually good practice that all public attributes for a non-Atom Basic class should be properties that access `args`. In some cases for performance reasons it can be better to attach some attributes directly on the instance but we still need to make sure that the func-args invariant is satisfied and that all of these attributes are correctly initialised by `obj.func(*obj.args)`.

Reply:
Well putting them in `_args` means, if I understand correctly, that it is possible to use `subs` to substitute these values, the question is whether that was intended. I'll be honest, this confuses me about these objects. Are they intended to behave like pure `Atom`s, that non of its arguments can be substituted, should it be a mix, that the arguments storted in _args can be substituted and those that are not in there can not, or was it intended that all arguments can be substituted, but the original author did not know how.

Reply:
I don't think we will get an answer to the question of what the original author intended so we should try now to decide what makes sense. @moorepants might be able to help with understanding this design.

Reply:
I think what makes most sense is 
> Is it not possible to change it so that the optional arguments for BaseScalar are passed to super().__new__ and appear in args?
Which is what I tried to accomplish with
```python
    def func(self, *args):
        return self.__new__(self.__class__, *args, pretty_str =
                self._pretty_form[1:], latex_str = self._latex_form[2:])
```
The slices are there because the `str` inputs get modified before they are stored.
This way, the non `args` arguments behave "atomic".

Reply:
Okay but can we make that work by defining `args` and `__new__` rather than `func`?

Reply:
I can make it work now, but these are subclasses of `Atom`, so they are expected to have no `args` in the first place and would fail if `Atom`s `func` returns `lambda: self`.

Reply:
But why are they subclasses of Atom? Maybe they shouldn't be. There is clearly something wrong here so we can't start from the "if it ain't broke don't fix it" perspective.

Reply:
The problems does remain though, if it works with `args` and `__new__` it always means that those arguments can be substituted as far as I understand. If you want any argument to behave "atomic", you will have to have some kind of "manipulation" in `func`

Reply:
It's not clear to me what the original authors intended. Few people understand how Basic works so few people understand how to subclass it correctly. What matters is not necessarily what those original authors intended but rather what is actually reasonable to us now and also whether any change is compatible with previous versions of SymPy.

Reply:
If it is required to be compatible with previous versions, then `pretty_str` and `latex_str` can not be stored in `_args` since that would make them substitutable. Then that only leaves the option of using a modified `func` so that these arguments are given to a newly created instance.

Currently, it seems to me that the "supported" way to store non-basic arguments (that then have the property that they can't be substituted or in general be modified with `obj.func( modification(obj.args) )`, but are still maintained under such operations) is to create a metaclass which stores those non-basic arguments as class attributes (e.g. I think `Function` does this with the name argument). This is needed, because you don't want the function to be replaced when you perform a sub with `obj.subs(Str('gamma'),Str('beta))` but only with `obj.subs(Function('gamma'),Function('beta'))`.

I think the lack of a generic way to introduce non `args` into `Basic`s and the (perceived) need to do so is what causes these kind of inconsistent codes. Something simple of demanding all `args` arguments to be first in `__new__` would work nicely, since `func` could then be written as:
```python
def __new__(self, a, b, c, d):
    obj = Basic.__new__(a, b)
    obj._non_args = (c, d)
def func(self, *args):
    return self.__class__(*args, *self._non_args)
```
This would work for `Atom`, `Basic` and "mixed" (with both arguments that behave "basic" and that behave "atomic"), although for performance reasons one might want to return self for `Atom` and leave out the `_non_args` in the case of `Basic`. (I do not suggest to add this in this PR or at all, it is just to indicate what I expected `Basic` to be able to do, and what I think also others expect it to do, or "feel" the need for `Basic` to do)

The proposed solution does exactly what is described above, except it isn't generic but specific for this class ( `*self._non_args` replaced with the actual arguments). 

I could work with any way to indicate an element of `args` is "atomic" (can't be replaced/modified). I could also make it work to simply treat these classes as `Atom` and don't store any of the arguments in `_args`. Which might actually be a good property for these classes.

Reply:
> If it is required to be compatible with previous versions, then `pretty_str` and `latex_str` can not be stored in `_args` since that would make them substitutable

It's not clear to me that substitutability matters in this situation or more importantly whether the current behaviour would be preferred over a more consistent alternative.

Reply:
Because when the intention is to use subs on a `Str` elsewhere in an `expr` also containing a `sympy.vector.scalar.BaseScalar` the `pretty_str` value might be replaced unintentionally simply because it coincidentally has the same `str` value, eventhough it represents a completely different kind of object.

Reply:
I suspect that that is very unlikely in practice so that it isn't really a compatibility problem. But that's why I brought @bjodah in because he might have more of an idea how this is used.

Reply:
I made an attempt at putting the strings in `args`. This leads to a bunch of test fails, which I think are due to the hash changing which changes the order in e.g. `Add._sorted_args`. Is it safe to assume that this change in behaviour will not cause issues? Or does this preclude the option of adding the strings to `args`?